### ISSUE TITLE
dropbox: uploading improvements

### ISSUE THREAD

**Issue by ncw** [rclone/rclone#5156]
Rclone's dropbox upload is not using several dropbox features which it should do to improve the upload speed and files/second.

- [Patching of uploads](https://developers.dropbox.com/dbx-performance-guide#batch-upload)
- Parallel chunked uploads for large files - [see upload session start](https://www.dropbox.com/developers/documentation/http/documentation#files-upload_session-start) and ` UploadSessionType.concurrent`

## Batching of uploads

This is well explained in the [performance guide](https://developers.dropbox.com/dbx-performance-guide#batch-upload) but to summarise, unless rclone uses batches to upload it is very likely to encounter `too_many_write_operations` or `too_many_requests` errors. This is very evident when trying to upload a directory of small files.

Batching uploads intersects awkwardly with rclone's architecture but it is possible.

## Parallel chunked uploads

Uploading large files with parallel chunks is a well trodden patch for rclone. Currently we use a 48MB chunk size. Suggestons are that a 16MiB chunk size with 4 concurrent uploads might work very well.

This needs to interact with the batching code.

## See also

- #4656 - PR to speedup chunked uploads
- #5008 - Reject attempts to upload larger objects

**Comment by ncw** [comment]
Here is the current iteration of the dropbox batch upload code

[v1.55.0-beta.5331.69b091f34.fix-dropbox-batch-sync](https://beta.rclone.org/branch/fix-dropbox-batch-sync/v1.55.0-beta.5331.69b091f34.fix-dropbox-batch-sync/) on branch [fix-dropbox-batch-sync](https://github.com/rclone/rclone/tree/fix-dropbox-batch-sync) (uploaded in 15-30 mins)

For a directory of 1000 5 byte files upload times are as follows

| batch_mode | batch_size | batch_timeout | --transfers | time |
| --- | --- | --- | --- | --- |
| off | - | - |  32 | 23m |
| sync | 32 | 0.5s | 32 | 4m48s |
| sync | 256 | 0.5s | 256 | 1m40s |
| async | 100 | 10s | 32 | 1m1s |
| async | 1000 | 10s | 32 | 23s |

So for transferring lots of small files at least `--dropbox-batch-mode` makes a huge difference. See the commit for discussion of what the parameters mean.

**Comment by ivandeex** [comment]
> Rclone's dropbox upload is not using several dropbox features which it should do to improve the upload speed and files/second.
> * [Patching of uploads](https://developers.dropbox.com/dbx-performance-guide#batch-upload)

Patching -> Batching?

**Comment by borjandev** [comment]
![Screenshot 2021-03-26 at 07 39 07](https://user-images.githubusercontent.com/64651636/112592500-5c096680-8e06-11eb-9524-70dd255cc540.png)

The linux workflow seems to have failed? https://github.com/rclone/rclone/runs/2176162859

Seems like the race test check didn't pass, any chance of restarting the build in order to get linux binaries?

Thanks!

**Comment by darthShadow** [comment]
Done. Builds are now available.

**Comment by borjandev** [comment]
Will `batch_mode` , `batch_size` and `batch_timeout` work with `rclone mount` as well on a dropbox remote? 

For example with `--vfs-cache-mode full` ?

The particular usecase would be transfering a bunch of small files from the local filesystem to the dropbox mount.

Thanks!

**Comment by ncw** [comment]
> Will `batch_mode` , `batch_size` and `batch_timeout` work with `rclone mount` as well on a dropbox remote?
> 
> For example with `--vfs-cache-mode full` ?

Yes they will. They may need some tuning with a mount but they will work just fine.
 
> The particular usecase would be transfering a bunch of small files from the local filesystem to the dropbox mount.

Perfect use of batchmode.

**Comment by NoLooseEnds** [comment]
I'm seeing this error with `rclone v1.55.0-beta.5331.69b091f34.fix-dropbox-batch-sync` on Linux, seems to be same issue as this [relatively recent post](https://forum.rclone.org/t/invalid-character-u-looking-for-beginning-of-value/20059).

```
2021/04/03 02:22:02 ERROR : /path/to/dir: error reading source directory: invalid character 'u' looking for beginning of value
2021/04/03 02:23:03 ERROR : path/to/dir:. Working with Subscriber Lists: error reading source directory: invalid character 'u' looking for beginning of value
2021/04/03 02:38:47 ERROR : path/to/dir:: error reading source directory: invalid character 'u' looking for beginning of value
```
Did not have the same problem with `rclone v1.54.0-beta.4758.c741c02fb.fix-dropbox-batch`

**Comment by dertasiu** [comment]
Hi! I've been trying it with the "serve" command because I normally use rclone as a backend of restic.
I don't see any "too_many_requests" errors anymore.
But the snapshot never finishes.
I've setup a restic repository for testing and I'm trying to upload a file with "async". At the beginning it gets pretty good speeds, but after a while, more like when it is finishing the upload, it stalls. The latest test was running for 33 minutes for a 700MB file.
I am attaching the output of restic with the output of rclone set to DEBUG.
[logfile.txt](https://github.com/rclone/rclone/files/6258069/logfile.txt)

**Comment by ncw** [comment]
@NoLooseEnds wrote:
> I'm seeing this error with `rclone v1.55.0-beta.5331.69b091f34.fix-dropbox-batch-sync` on Linux, seems to be same issue as this [relatively recent post](https://forum.rclone.org/t/invalid-character-u-looking-for-beginning-of-value/20059).
> 
> ```
> 2021/04/03 02:22:02 ERROR : path/to/dir: error reading source directory: invalid character 'u' looking for beginning of value
> 2021/04/03 02:23:03 ERROR : path/to/dir:. Working with Subscriber Lists: error reading source directory: invalid character 'u' looking for beginning of value
> 2021/04/03 02:38:47 ERROR : path/to/dir:: error reading source directory: invalid character 'u' looking for beginning of value
> ```
> 
> Did not have the same problem with `rclone v1.54.0-beta.4758.c741c02fb.fix-dropbox-batch`

Can you run with `-vv --dump responses` and attach a log that would be fantastic.

**Comment by ncw** [comment]
@dertasiu wrote:
> Hi! I've been trying it with the "serve" command because I normally use rclone as a backend of restic.
> I don't see any "too_many_requests" errors anymore.
> But the snapshot never finishes.
> I've setup a restic repository for testing and I'm trying to upload a file with "async". At the beginning it gets pretty good speeds, but after a while, more like when it is finishing the upload, it stalls. The latest test was running for 33 minutes for a 700MB file.
> I am attaching the output of restic with the output of rclone set to DEBUG.
> [logfile.txt](https://github.com/rclone/rclone/files/6258069/logfile.txt)

Hmm, it looks like the batch upload finished and then deadlocked.

Can you `kill -QUIT` the rclone process when it is stuck and attach the backtrace here please?

**Comment by dertasiu** [comment]
@ncw wrote:
> Hmm, it looks like the batch upload finished and then deadlocked.
> 
> Can you `kill -QUIT` the rclone process when it is stuck and attach the backtrace here please?

Of course!:
[rclone_QUIT_log.txt](https://github.com/rclone/rclone/files/6260326/rclone_QUIT_log.txt)

**Comment by NoLooseEnds** [comment]
@ncw Ok, reporting back. 

I ran the exact same command, except for turning on `debug` logging and `--dump responses`, and this time it went through fine. Same dataset (without changes), same everything. 

So except for a humungous log file, all is good. I'll report back if it occurs again (it's a script that runs daily). 

The command for reference:
```
rclone sync dropbox:/ gdrive-crypt-sa:/last_snapshot --backup-dir=gdrive-crypt-sa:/archive/2021/2021-04-05_20:47:46 --log-file=/home/<user>/rclone/log/rclone_jobber.log --rc --rc-no-auth --config /home/<user>/rclone/rclone.conf --filter-from /home/<user>/rclone/filter_rules --fast-list --transfers 24 --checkers 24 --dropbox-chunk-size 150M --dropbox-batch-mode sync --tpslimit 12 --tpslimit-burst 12 --drive-chunk-size 128M --create-empty-src-dirs --log-level DEBUG --dump responses
```

**Comment by ncw** [comment]
> > Hmm, it looks like the batch upload finished and then deadlocked.
> > Can you `kill -QUIT` the rclone process when it is stuck and attach the backtrace here please?
> 
> Of course!:
> [rclone_QUIT_log.txt](https://github.com/rclone/rclone/files/6260326/rclone_QUIT_log.txt)

I can see what happened in that traceback - thank you!

There was a bit of dead code in the commit handler from a previous iteration which was eating the commits and causing a deadlock. Stupid now I look at it!

This should fix it

[v1.55.0-beta.5332.d732cf2e9.fix-dropbox-batch-sync](https://beta.rclone.org/branch/fix-dropbox-batch-sync/v1.55.0-beta.5332.d732cf2e9.fix-dropbox-batch-sync/) on branch [fix-dropbox-batch-sync](https://github.com/rclone/rclone/tree/fix-dropbox-batch-sync) (uploaded in 15-30 mins)


> I ran the exact same command, except for turning on `debug` logging and `--dump responses`, and this time it worked fine. Same dataset (without changes), same everything. Not sure what caused it, but I'm guessing something on dropbox's side.

The bug I fixed would be somewhat non-deterministic so you might not see it every time. Though it did happen 5 times in the log you sent :shrug:

**Comment by dertasiu** [comment]
> > > Hmm, it looks like the batch upload finished and then deadlocked.
> > > Can you `kill -QUIT` the rclone process when it is stuck and attach the backtrace here please?
> > 
> > 
> > Of course!:
> > [rclone_QUIT_log.txt](https://github.com/rclone/rclone/files/6260326/rclone_QUIT_log.txt)
> 
> I can see what happened in that traceback - thank you!
> 
> There was a bit of dead code in the commit handler from a previous iteration which was eating the commits and causing a deadlock. Stupid now I look at it!
> 
> This should fix it
> 
> [v1.55.0-beta.5332.d732cf2e9.fix-dropbox-batch-sync](https://beta.rclone.org/branch/fix-dropbox-batch-sync/v1.55.0-beta.5332.d732cf2e9.fix-dropbox-batch-sync/) on branch [fix-dropbox-batch-sync](https://github.com/rclone/rclone/tree/fix-dropbox-batch-sync) (uploaded in 15-30 mins)

Wow, it works perfectly now.
Thank you very much!!

**Comment by NoLooseEnds** [comment]
@ncw Seems like the latest beta is missing a Linux build again.

**Comment by ncw** [comment]
I gave the build a poke - hopefully it will build this time @NoLooseEnds

**Comment by ncw** [comment]
Here is a new build which tries harder to commit the batch

[v1.55.0-beta.5333.d65916f9a.fix-dropbox-batch-sync](https://beta.rclone.org/branch/fix-dropbox-batch-sync/v1.55.0-beta.5333.d65916f9a.fix-dropbox-batch-sync/) on branch [fix-dropbox-batch-sync](https://github.com/rclone/rclone/tree/fix-dropbox-batch-sync) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
Here is a new build which fixes a bug with not committing the last batch when using batch mode async.

[v1.55.0-beta.5334.ac7acb9f0.fix-dropbox-batch-sync](https://beta.rclone.org/branch/fix-dropbox-batch-sync/v1.55.0-beta.5334.ac7acb9f0.fix-dropbox-batch-sync/) on branch [fix-dropbox-batch-sync](https://github.com/rclone/rclone/tree/fix-dropbox-batch-sync) (uploaded in 15-30 mins)

**Comment by mauricev** [comment]
I'm testing with v1.56.0-beta.5420.2a40f0007 on Windows 10, uploading a 20 GB file. 

The performance is about 15 MB/second, which isn't great, and it repeatedly stalls out every so often (about every 20 chunks or so). When I hit "return", that wakes it up and continues until the next stall. 

I get an unknown flag error with --dropbox-batch-mode.

**Comment by PatrickHuetter** [comment]
@mauricev you aren't using the correct beta version from the fix-dropbox-batch-sync branch. So the batch mode improvements aren't available for you and you're working with the old algorithm/api.

**Comment by mauricev** [comment]
That version is even worse than the standard version, with even more stalling and worse performance.

**Comment by PatrickHuetter** [comment]
> That version is even worse than the standard version, with even more stalling and worse performance.

I can't confirm that. For me it was a huge improvement. Uploading a lot of small files ~4TB wasn't possible with the old version. With this improvements here i was able to upload everything. I've used the async mode and checked after sync one more time.

**Comment by nh2** [comment]
I'm trying with the latest commit https://github.com/rclone/rclone/commit/d167025660e1caec5deb6bc31f6a6e98bf50fa1a on branch [`fix-dropbox-batch-sync`](https://github.com/rclone/rclone/compare/fix-dropbox-batch-sync) with:

```sh
rclone sync ... --checkers 32 transfers 32 --dropbox-batch-size 32 --dropbox-batch-mode async
```

Trying to upload a directory with 2TB and 3M files, I get:

```
2021-05-02 19:34:24 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021-05-02 19:34:24 NOTICE: too_many_requests/..: Too many requests or write operations. Trying again in 300 seconds.
2021-05-02 19:34:24 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021-05-02 19:34:24 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021-05-02 19:34:24 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2021-05-02 19:34:24 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
```

Upload speed is `1.9 MBytes/s` -- very low for this machine that has a 10 Gbit/s and can upload up to 500 MB/s when uploading larger files to Dropbox with `rclone sync` (without `--dropbox-batch-mode`).

**Comment by NoLooseEnds** [comment]
@nh2 Try adding  `--tpslimit 12 --tpslimit-burst 12`

I got rate limited by dropbox, and experimented until I found the sweet spot. It's been a while now, and I have not tweaked in a while (i.e don't know I dropbox has change anything).

**Comment by ncw** [comment]
@nh2 wrote:
> I'm trying with the latest commit [d167025](https://github.com/rclone/rclone/commit/d167025660e1caec5deb6bc31f6a6e98bf50fa1a) on branch [`fix-dropbox-batch-sync`](https://github.com/rclone/rclone/compare/fix-dropbox-batch-sync) with:
> 
> ```shell
> rclone sync ... --checkers 32 transfers 32 --dropbox-batch-size 32 --dropbox-batch-mode async
> ```
> 
> Trying to upload a directory with 2TB and 3M files, I get:
> 
> ```
> 2021-05-02 19:34:24 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
> 2021-05-02 19:34:24 NOTICE: too_many_requests/..: Too many requests or write operations. Trying again in 300 seconds.
> 2021-05-02 19:34:24 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
> 2021-05-02 19:34:24 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
> 2021-05-02 19:34:24 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
> 2021-05-02 19:34:24 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
> ```

We shouldn't be getting those errors with the batching code.

> Upload speed is `1.9 MBytes/s` -- very low for this machine that has a 10 Gbit/s and can upload up to 500 MB/s when uploading larger files to Dropbox with `rclone sync` (without `--dropbox-batch-mode`).

Those 300 second delays are killing the performance!

Are you running more than one copy of rclone at once uploading things? Dropbox really doesn't like that.

@NoLooseEnds wrote
> Try adding `--tpslimit 12 --tpslimit-burst 12`
> 
> I got rate limited by Dropbox and experimented until I found the sweet spot. It's been a while now, and I have not tweaked in a while (i.e don't know I dropbox has changed anything).

Are you running multiple rclones at once?

**Comment by nh2** [comment]
> Are you running more than one copy of rclone at once uploading things? Dropbox really doesn't like that.

No; I did have a normal-rclone-version running, but I Ctrl+Z'ed it before starting the run from the `fix-dropbox-batch-sync` branch.

I also have `rclone mount` running against it in the background, but no reads or writes are happening to that mount.

**Comment by NoLooseEnds** [comment]
@ncw No. Only one sync instance. 

We (through PMs on rclone forum) experimented with that tpslimit on an earlier dropbox build (rclone v1.54.0-beta.4758.c741c02fb.fix-dropbox-batch).

**Comment by nh2** [comment]
@ncw I have now killed all previous `rclone sync`s, waited 5 minutes to ensure that all 300 second timeouts are over, and restarted with `-vv --dump responses`.

I still get very low throughput, and few requests per second: I'm at `xfr#12526/23039` after 80 minutes, making it ~2.5 requests per second.

I also do get `too_many_requests` responses; here is an excerpt:

```
2021-05-03 23:59:56 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2021-05-03 23:59:56 NOTICE: too_many_requests/..: Too many requests or write operations. Trying again in 300 seconds.
2021-05-03 23:59:56 DEBUG : pacer: low level retry 1/10 (error too_many_requests/..)
2021-05-03 23:59:56 DEBUG : pacer: Rate limited, increasing sleep to 5m0s
2021-05-03 23:59:56 DEBUG : data/00f6385e159b923a2969d78125dfe574b3c313fa4ecb2d9ea902a4a1c2532233.bin: Uploading chunk 1/1
2021-05-03 23:59:56 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2021-05-03 23:59:56 DEBUG : HTTP RESPONSE (req 0xc04bac7f00)
2021-05-03 23:59:56 DEBUG : HTTP/1.1 200 OK
Transfer-Encoding: chunked
Cache-Control: no-cache
Content-Type: application/json
Date: Mon, 03 May 2021 23:59:56 GMT
Pragma: no-cache
Server: envoy
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Dropbox-Request-Id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
X-Dropbox-Response-Origin: remote
X-Robots-Tag: noindex, nofollow, noimageindex
X-Robots-Tag: noindex, nofollow, noimageindex
X-Server-Response-Time: 1820

4
null
0
2021-05-03 23:59:56 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2021-05-03 23:59:56 DEBUG : pacer: Reducing sleep to 3m45s
2021-05-03 23:59:56 DEBUG : data/00f5ef2a9513233acdbf233e6f683b329abbf4e9cc2cd9a7b3e23fd410abfa48.bin: Uploading chunk 2/1
2021-05-03 23:59:56 DEBUG : <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<
2021-05-03 23:59:56 DEBUG : HTTP RESPONSE (req 0xc050813600)
2021-05-03 23:59:56 DEBUG : HTTP/1.1 429 Too Many Requests
Transfer-Encoding: chunked
Cache-Control: no-cache
Content-Security-Policy: sandbox allow-forms allow-scripts
Content-Type: application/json
Date: Mon, 03 May 2021 23:59:56 GMT
Retry-After: 300
Server: envoy
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Dropbox-Request-Id: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
X-Dropbox-Response-Origin: remote
X-Robots-Tag: noindex, nofollow, noimageindex
X-Robots-Tag: noindex, nofollow, noimageindex

70
{"error_summary": "too_many_requests/.", "error": {"reason": {".tag": "too_many_requests"}, "retry_after": 300}}
```

**Comment by ncw** [comment]
Acording to the docs: https://developers.dropbox.com/dbx-performance-guide

"too_many_requests" is a rate limit error. Empirically other users have found that `--tpslimit 12` always stays under the dropbox rate limit - can you give that a go?

Also `--transfers 512` sound a little high - this is making 512 connections to dropbox at once! So I'd be inclined to lower it to 64 or so.

**Comment by PatrickHuetter** [comment]
Okay i also have the `NOTICE: too_many_write_operations/: Too many requests or write operations. Trying again in 1 seconds.` now. I think there might be a problem with some temporary files that get often written in the rclone mount. There is another ticket about this but i think an exlude filter for mount (folder and file based) would help for this.

I only have one rclone stable (1.55.1) instance running with my own custom oauth2 app created in dropbox backend.
I also tested the tpslimits and all other settings. Nothing helped. 

@ncw If you want, we can exchange contact data and have a google meet/hangout. I can show you live on my screen the problems. We can go through it. It must be some small issue because i already tested and isolated everything.

**Comment by nh2** [comment]
> Also `--transfers 512` sound a little high - this is making 512 connections to dropbox at once! So I'd be inclined to lower it to 64 or so.

@ncw I may then not have fully understood how the options `--dropbox-batch-size` and `--transfers` interact with each other. Do they multiply (e.g. _each_ of the connections will issue batches of size `dropbox-batch-size`)? My understanding was no, because the added docs say:

> `--dropbox-batch-mode sync`
>
> In this mode rclone will batch up uploads to the size specified by `--dropbox-batch-size` and commit them together.
>
> Using this mode means you can use a much higher `--transfers` parameter (32 or 64 works fine) without receiving `too_many_requests` errors.

So I was under the impression that if I want to make use of a large batch size, I needed to set `--transfers` equally high. Otherwise, if they do multiply, I didn't quite understand why in your benchmark in https://github.com/rclone/rclone/issues/5156#issuecomment-804968043 you had the combination of batch size 256 and transfers 256, given that there were only 1000 files.

And from the linked Dropbox docs I understood that 512 was still well below the maximum batch limit:

> Group files that are being uploaded into one or more batches. A batch can contain up to 1,000 entries.

So I thought that the most efficient way to use it (least requests) would be to make the batch size as close to 1000 as possible.

Please let me know which of this is right.

**Comment by nh2** [comment]
I'm now trying `--checkers 32 --transfers 32 --dropbox-batch-size 512 --dropbox-batch-mode sync`, with and without `--tpslimit 12`, but am not having any luck to get more than ~2 files transmitted per second.

Without `--tpslimit 12`, there was a single `too_many_requests`; with it there were none.

**Comment by nh2** [comment]
In the `sync` that I've been running for the last couple days which is running very slowly (by now the average speed has dropped to `107 kByte/s` on this 10 Gbit/s machine), these few errors are the only output:

```
2021-05-05 07:19:38 ERROR : Dropbox root 'myrepo': sync batch commit: failed to commit batch length 29: wait for batch failed after 120 tries: batch didn't complete
2021-05-05 10:33:52 ERROR : Dropbox root 'myrepo': sync batch commit: failed to commit batch length 1: wait for batch failed after 120 tries: internal_error/..
2021-05-05 13:18:43 ERROR : Dropbox root 'myrepo': sync batch commit: failed to commit batch length 1: wait for batch failed after 120 tries: internal_error/...
2021-05-06 00:13:00 ERROR : Dropbox root 'myrepo': sync batch commit: failed to commit batch length 1: wait for batch failed after 120 tries: batch didn't complete
```

**Comment by ncw** [comment]
I can make the max time to wait for a batch configurable - do you think that is worth doing @nh2?

Why is the sync running so slowly? Are you getting lots of retry after errors?

**Comment by nh2** [comment]
@ncw I do not know much yet about rclone internals, but I have the suspicion that something may be not right yet in general; with the branch so far I have never observed more than 1-2 operations per second, be it via `sync` or via `mount`.

> Are you getting lots of retry after errors?

The above errors is the only output I get (in addition to the progress line).

**Comment by ncw** [comment]
I've merged the state of play so far into master so this can get more testing before the release. See [the latest beta](https://beta.rclone.org/) (in 30 mins from now).

@nh2 so you aren't seeing any `NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.` errors?

I will put the configurable timeout in so you can test that next week.

**Comment by nh2** [comment]
> so you aren't seeing any `NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.` errors?

@ncw That is correct, I posted all output. The only remainder is the updating status line at the bottom:

```
2021/05/15 15:56:59 -    48.647G / 58.877 GBytes, 83%, 55.573 kBytes/s, ETA 2d5h37m3s (xfr#48307/58340)
```

(Note the `xfr` line is misleading; I have millions of files in, but I believe rclone discoveres them on the fly.)

**Comment by ncw** [comment]
@nh2 I can only assume that dropbox is throttling the file creation rate through delaying the batch completions then...

**Comment by NoLooseEnds** [comment]
I started seeing this today, so I'm going to update to the latest beta and see if that changes anything. Could also be a hiccup.

Currently running:
```
rclone sync dropbox:/ gdrive-crypt-sa:/last_snapshot --backup-dir=gdrive-crypt-sa:/archive/2021/2021-06-15_18:00:01 --log-file=/home/<user>/rclone/log/rclone_jobber.log --rc --rc-no-auth --config /home/<user>/rclone/rclone.conf --filter-from /home/<user>/rclone/filter_rules --fast-list --transfers 24 --checkers 24 --dropbox-chunk-size 150M --dropbox-batch-mode sync --tpslimit 12 --tpslimit-burst 12 --drive-chunk-size 128M --create-empty-src-dirs --user-agent PqJiMsNL2pkPYdwwqUR7ttBYUuTnPDz4Pt7gkNACs --log-level INFO
```


```
rclone v1.56.0-beta.5394.d16702566.fix-dropbox-batch-sync
- os/version: debian 10.9 (64 bit)
- os/kernel: 4.19.0-16-amd64 (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.16.3
- go/linking: static
- go/tags: none
```

```
2021/06/16 06:59:37 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:37 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:37 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:37 NOTICE: too_many_requests/..: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:37 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:37 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:38 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:39 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:39 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:39 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:39 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 06:59:39 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2021/06/16 07:00:05 INFO  :
Transferred:      498.537M / 498.537 MBytes, 100%, 2.876 MBytes/s, ETA 0s
Checks:           1919623 / 1919623, 100%
Deleted:                2 (files), 0 (dirs)
Renamed:                5
Transferred:           10 / 10, 100%
Elapsed time:    13h0m2.3s
```

**Comment by ncw** [comment]
> Wouldn't it be possible to implement a logic that limits itself based on rate limits, i.e. tries different tpslimit? I notice even when we upload a lot of small files (even in batch) that we often run into "too_many_write_operations" errors.

That should be what is happening. Rclone will back off when it gets those `too_many_write_operations` errors and it will obey any timeouts that dropbox sends. Sometimes those timeouts can be very long though.

The non-JSON error should never happen and I'd class as a dropbox bug.

**Comment by NoLooseEnds** [comment]
This was a bit of a strange experience, that makes little sense.

1. First I lowered the `--tpslimit` and `--tpslimit-burst` step by step, as far down as 1. Still did not work, I still got the error, and everything took ages.
2. Then I suddenly started to [get errors](https://forum.rclone.org/t/suddenly-getting-googleapi-error-401-invalid-credentials-autherror/27233) `googleapi: Error 401: Invalid Credentials, authError` on some files out of the blue

The only change I did was the `tps-limit`, which should not affect google drive afaik. I tried everything, but the error kept coming. Finally, after a few days, it just started working again (except it took ages because of `tps-limit 1`). And then I tried increasing it again to `12`, and it worked.

Now everything is back exactly like it was before I got the `error reading source directory: invalid character '<' looking for beginning of value` dropbox error. And the Google auth error is also gone.

So no idea what was causing this. Came out of nowhere, and started working again out of nowhere

**Comment by ivandeex** [comment]
@NoLooseEnds 

Google auth? I thought this ticket spins around Dropbox. :)

**Comment by NoLooseEnds** [comment]
@ivandeex 
Not relevant to this issue at all, it's more anecdotal.
Two strange issues occurred out of the blue, nothing worked to fix it, until it suddenly just worked again (exactly as it was before).

**Comment by animosity22** [comment]
I was going through some of this since I've migrated over to dropbox and I was just testing out the limits as well and 12 seemed to be a sweet spot but for some things, I don't get how the upload changes are impacted other operations.

If a run a size for example, it tags the committing uploads message at the end and 12 works, while 15 dies.

```
etexter@seraphite ~ % rclone size DB: -vvv --tpslimit 12 --tpslimit-burst 12
2021/12/08 07:48:33 INFO  : Starting transaction limiter: max 12 transactions/s with burst 12
2021/12/08 07:48:33 DEBUG : rclone: Version "v1.57.0" starting with parameters ["rclone" "size" "DB:" "-vvv" "--tpslimit" "12" "--tpslimit-burst" "12"]
2021/12/08 07:48:33 DEBUG : Creating backend with remote "DB:"
2021/12/08 07:48:33 DEBUG : Using config file from "/Users/etexter/.config/rclone/rclone.conf"
Total objects: 55.699k (55699)
Total size: 180.908 TiB (198910637106946 Byte)
2021/12/08 07:56:26 DEBUG : 7 go routines active
2021/12/08 07:56:26 INFO  : Dropbox root '': Commiting uploads - please wait...
etexter@seraphite ~ % rclone size DB: --tpslimit 15 --tpslimit-burst 15
2021/12/08 08:04:48 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
2021/12/08 08:04:48 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/12/08 08:04:48 NOTICE: too_many_requests/...: Too many requests or write operations. Trying again in 300 seconds.
2021/12/08 08:04:48 NOTICE: too_many_requests/.: Too many requests or write operations. Trying again in 300 seconds.
```

For what it's worth when I was migrating data, I had 3 gigabit machines going concurrently for days and rarely every saw an issue and I had no limits anywhere and was moving ~20 transfers at a time but it was coming from the Drive API which has a limiter on it by default so wasn't sure if that was hiding the issue.

Is there a benefit to adding some default pacer logic like with Google if there is a 'hidden' limit of 12 per second? I did push Dropbox on the API Quotas but they didn't share any specifics.
I have a spare account since the limits are per user based so I can test pretty easily without breaking things

**Comment by dennisoderwald** [comment]
We now work with a tpslimit of 10 per Dropbox app. Anything above that we have encountered problems regardless of the anz. Transfers and/or sync/async.

But you can immediately bypass the limits with a new token.

**Comment by dennisoderwald** [comment]
We now work with a tpslimit of 10 per Dropbox app. Anything above that we have encountered problems regardless of the number of Transfers and/or sync/async.

But you can immediately bypass the limits with a new token. And yes, Dropbox itself does not comment on this.

With big files it works more, but with small files it doesn't work more than 10. Commands like "size" are very resource hungry.

**Comment by animosity22** [comment]
From my chat with support, the limits were per user and you could have many apps but if they all tied back to the same user, you'd be in the same boat. My apps are not production either as I'm not sure if that matters or not.

**Comment by dennisoderwald** [comment]
We don't have any production applications but have validated exactly what the support also told me and in our account it is definitely different.

We have several dev scoped applications and the limits are related to the application and the token.

**Comment by animosity22** [comment]
From my support email:

```
The Dropbox API does have a rate limiting system, but we don't have any specific numbers documented. It is only designed to prevent abuse though, and is accordingly very generous. Further, the limits operate on a per-user basis. That being the case, you generally don't need to worry about hitting it in normal use. The Dropbox API rate limiting system operates the same regardless of account type.

```

**Comment by dennisoderwald** [comment]
I actually believe that they are not talking about the "too_many_requests" error with the rate limit.

**Comment by animosity22** [comment]
They were as that was my specific ask:

![image](https://user-images.githubusercontent.com/5088712/145221178-127f268f-1267-4aeb-9216-757323769ebd.png)

**Comment by animosity22** [comment]
I did test what you were saying and think your support guy was correct as I have 2 Apps on my one user. I am able to trigger a rate failure with just some size commands and everything fails against that app and the same user on a different app works. Appreciate talking through that as it's helpful to get a 'right' answer as the 1st level support seems to be reading from a script :(


```
etexter@seraphite rclone % rclone about DB2:
Total:   198.215 TiB
Used:    180.908 TiB
Free:    17.307 TiB
etexter@seraphite rclone % rclone about DB:
2021/12/08 10:15:31 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
2021/12/08 10:15:31 NOTICE: too_many_requests/: Too many requests or write operations. Trying again in 300 seconds.
```

So for me, I did a size command on DB: to trigger a rate limit and as expected, all commands fail and the same user on the 2nd app I setup, works without issue.

**Comment by dennisoderwald** [comment]
I can only tell you from experience that the support usually (unfortunately) has no idea. I have also received the statement that "rclone" would not work at all but only official clients, that rclone only uses the API, the person did not care.

What I want to say: I would not take the statements of Dropbox for full. Or did you talk to the API team?

**Comment by animosity22** [comment]
I spoke(emailed) withe the API team and didn't mention rclone at all.

I tested again with one user and validated quite a number of times it's per app registration so that was helpful as I have my setup a bit more segmented and have proper limits in place and haven't had an issue since.

**Comment by dennisoderwald** [comment]
What parameters are you running it on now?

**Comment by animosity22** [comment]
I've followed the advice from this item and limit my tps to 12:

```
--tpslimit 12 --tpslimit-burst 12
```

I ran through multiple iterations with just doing rclone size and limiting things as anything above 12 seemed to consistently break for me. I tested with a standalone app registration so I'm sure that was the only thing hitting the API.

```
etexter@seraphite rclone % rclone size DB: --tpslimit 12 --tpslimit-burst 12
Total objects: 55.210k (55210)
Total size: 182.300 TiB (200440950254545 Byte)
```

Anything above that 12 would rate limit me every time and those are only listing requests and no download/upload so it's a pretty repeatable test that doesn't depend on bandwidth/transfer speeds/etc.

**Comment by Rootax** [comment]
Hello,

"Parallel chunked uploads for large files " => Is it still being working on ?

Thx.

**Comment by ncw** [comment]
@Rootax wrote
> "Parallel chunked uploads for large files " => Is it still being working on ?

You are right we never implemented that.

That certainly could be implemented though - I don't think it would be too hard. Do you want to have a go @Rootax ?

**Comment by Rootax** [comment]
Could this be on the roadmap for 1.64 ?

**Comment by Rootax** [comment]
Sorry to dig this, but is this function in rclone now ?

Thx a lot :)

**Comment by ncw** [comment]
@Rootax rclone doesn't do parallel chunked uploads to Dropbox yet ( I double checked the code!)

We should implement OpenChunkedWriter for Dropbox which would be quite easy and that would enable parallel uploads and even parallel copies from other backend.

Want to have a go?
